File: dart/blockchain/lib/consensus/interpreters/eta_calculation.dart
import 'dart:typed_data';

import 'package:blockchain/codecs.dart';
import 'package:blockchain/common/algebras/clock_algebra.dart';
import 'package:blockchain/common/utils.dart';
import 'package:blockchain/common/models/common.dart';
import 'package:blockchain/consensus/algebras/eta_calculation_algebra.dart';
import 'package:blockchain/consensus/utils.dart';
import 'package:logging/logging.dart';
import 'package:blockchain_protobuf/models/core.pb.dart';

import 'package:fixnum/fixnum.dart';
import 'package:hashlib/hashlib.dart';

class EtaCalculation extends EtaCalculationAlgebra {
  final Future<SlotData> Function(BlockId) fetchSlotData;
  final ClockAlgebra clock;
  final Eta genesisEta;

  EtaCalculation(this.fetchSlotData, this.clock, this.genesisEta);

  final log = Logger("EtaCalculation");

  @override
  Future<Eta> etaToBe(SlotId parentSlotId, Int64 childSlot) async {
    if (childSlot > clock.slotsPerEpoch) return genesisEta;
    final parentEpoch = clock.epochOfSlot(parentSlotId.slot);
    final childEpoch = clock.epochOfSlot(childSlot);
    final parentSlotData = await fetchSlotData(parentSlotId.blockId);
    if (parentEpoch == childEpoch)
      return parentSlotData.eta;
    else if (childEpoch - parentEpoch > 1)
      throw Exception("Empty Epoch");
    else
      return _calculate(await _locateTwoThirdsBest(parentSlotData));
  }

  _locateTwoThirdsBest(SlotData from) async {
    if (_isWithinTwoThirds(from))
      return from;
    else
      return _locateTwoThirdsBest(
          await fetchSlotData(from.parentSlotId.blockId));
  }

  _isWithinTwoThirds(SlotData from) =>
      from.slotId.slot % clock.slotsPerEpoch <= (clock.slotsPerEpoch * 2 ~/ 3);

  Future<Eta> _calculate(SlotData twoThirdsBest) async {
    // TODO: Caching
    final epoch = clock.epochOfSlot(twoThirdsBest.slotId.slot);
    final epochRange = clock.epochRange(epoch);
    final epochData = [twoThirdsBest];
    while (epochData.first.parentSlotId.slot >= epochRange.$1) {
      epochData.insert(
          0, await fetchSlotData(epochData.first.parentSlotId.blockId));
    }
    final rhoValues = epochData.map((slotData) => slotData.rho);
    return _calculateFromValues(twoThirdsBest.eta, epoch + 1, rhoValues);
  }

  Eta _calculateFromValues(
      Eta previousEta, Int64 epoch, Iterable<Rho> rhoValues) {
    final rhoNonceHashValues = rhoValues.map((rho) => rho.rhoNonceHash);
    final args = EtaCalculationArgs(previousEta, epoch, rhoNonceHashValues);
    final eta = args.eta;
    log.info(
        "Calculated eta for epoch=$epoch eta=${eta.show} previousEta=${previousEta.show}");

    return eta;
  }
}

class EtaCalculationArgs {
  final Eta previousEta;
  final Int64 epoch;
  final Iterable<List<int>> rhoNonceHashValues;

  EtaCalculationArgs(this.previousEta, this.epoch, this.rhoNonceHashValues);

  Uint8List get eta {
    final bytes = <int>[]
      ..addAll(previousEta)
      ..addAll(epoch.toBigInt.bytes);
    rhoNonceHashValues.forEach(bytes.addAll);

    return blake2b256.convert(bytes).bytes;
  }
}

File: scala/core/src/main/scala/blockchain/consensus/EtaCalculation.scala
package blockchain.consensus

import blockchain.*
import blockchain.codecs.{*, given}
import blockchain.crypto.*
import blockchain.models.{BlockHeader, BlockId, SlotId}
import blockchain.utility.*
import cats.*
import cats.data.*
import cats.effect.{Clock as _, *}
import cats.implicits.*
import com.github.benmanes.caffeine.cache.Caffeine
import com.google.protobuf.ByteString
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.Slf4jLogger
import scalacache.caffeine.CaffeineCache
import scalacache.{Cache, Entry}

trait EtaCalculation[F[_]]:
  def etaToBe(parentSlotId: SlotId, childSlot: Long): F[Eta]

object EtaCalculation:
  def make[F[_]: Async: Parallel: CryptoResources](
      fetchHeader: BlockId => F[BlockHeader],
      clock: Clock[F],
      genesisEta: Eta
  ): Resource[F, EtaCalculation[F]] =
    Resource.pure(
      new EtaCalculationImpl[F](
        fetchHeader,
        clock,
        genesisEta,
        CaffeineCache[F, BlockId, Eta](
          Caffeine.newBuilder.maximumSize(32).build[BlockId, Entry[Eta]]()
        )
      )
    )

private[consensus] class EtaCalculationImpl[F[_]: Async: Parallel: CryptoResources](
    fetchHeader: BlockId => F[BlockHeader],
    clock: Clock[F],
    genesisEta: Eta,
    cache: Cache[F, BlockId, Eta]
) extends EtaCalculation[F] {

  implicit private val logger: Logger[F] =
    Slf4jLogger.getLoggerFromName("Blockchain.Consensus.EtaCalculation")

  private val twoThirdsLength = clock.epochLength * 2 / 3

  override def etaToBe(parentSlotId: SlotId, childSlot: Long): F[Eta] =
    clock
      .epochOf(childSlot)
      .flatMap(childEpoch =>
        if (childEpoch === 0L) genesisEta.pure[F]
        else
          for {
            parentEpoch <- clock.epochOf(parentSlotId.slot)
            parentSlotData <- fetchHeader(parentSlotId.blockId)
            eta <-
              if (parentEpoch === childEpoch)
                Async[F].delay(parentSlotData.eligibilityCertificate.eta.decodeBase58)
              else if (childEpoch - parentEpoch > 1)
                MonadThrow[F].raiseError(
                  new IllegalStateException(
                    show"Eta calculation encountered empty epoch for" +
                      show" parentSlotId=$parentSlotId" +
                      show" childSlot=$childSlot" +
                      show" parentEpoch=$parentEpoch" +
                      show" childEpoch=$childEpoch"
                  )
                )
              else locateTwoThirdsBest(parentSlotData).flatMap(calculate)
          } yield eta
      )

  /** Given some header near the end of an epoch, traverse the chain (toward genesis) until reaching a block that is
    * inside of the 2/3 window of the epoch
    */
  private def locateTwoThirdsBest(from: BlockHeader): F[BlockHeader] =
    if (isWithinTwoThirds(from)) from.pure[F]
    else
      from.iterateUntilM(data => fetchHeader(data.parentHeaderId))(
        isWithinTwoThirds
      )

  private def isWithinTwoThirds(from: BlockHeader): Boolean =
    from.slot % clock.epochLength <= twoThirdsLength

  /** Compute the Eta value for the epoch containing the given header
    * @param twoThirdsBest
    *   The latest block header in some tine, but within the first 2/3 of the epoch
    */
  private def calculate(twoThirdsBest: BlockHeader): F[Eta] =
    cache.cachingF(twoThirdsBest.id)(ttl = None)(
      Sync[F].defer(
        for {
          epoch <- clock.epochOf(twoThirdsBest.slot)
          epochRange <- clock.epochRange(epoch)
          epochData <- NonEmptyChain(twoThirdsBest).iterateUntilM(items =>
            fetchHeader(items.head.parentHeaderId).map(items.prepend)
          )(items => items.head.parentSlot < epochRange.start)
          rhoValues <- epochData.parTraverse(header => CryptoResources[F].ed25519VRF.useSync(e => header.rho(using e)))
          nextEta <- calculate(
            previousEta = twoThirdsBest.eligibilityCertificate.eta.decodeBase58,
            epoch = epoch + 1,
            rhoValues = rhoValues
          )
        } yield nextEta
      )
    )

  /** Calculate a new Eta value once all the necessary pre-requisites have been gathered
    */
  private def calculate(
      previousEta: Eta,
      epoch: Long,
      rhoValues: NonEmptyChain[Rho]
  ): F[Eta] =
    for {
      _ <- Logger[F].debug(
        show"Calculating new eta." +
          show" previousEta=$previousEta" +
          show" epoch=$epoch" +
          show" rhoValues=[${rhoValues.length}]{${rhoValues.head}..${rhoValues.last}}"
      )
      rhoNonceHashes <- rhoValues
        .parTraverse(rho =>
          CryptoResources[F].blake2b512
            .useSync(implicit b2b => rhoToRhoNonceHash(rho))
        )
      nextEta <- calculateFromNonceHashValues(
        previousEta,
        epoch,
        rhoNonceHashes
      )
      _ <- Logger[F].info(
        show"Calculated new eta." +
          show" previousEta=$previousEta" +
          show" epoch=$epoch" +
          show" rhoValues=[${rhoValues.length}]{${rhoValues.head}..${rhoValues.last}}" +
          show" nextEta=$nextEta"
      )
    } yield nextEta

  /** Calculate a new Eta value once all the necessary pre-requisites have been gathered
    */
  private def calculateFromNonceHashValues(
      previousEta: Eta,
      epoch: Long,
      rhoNonceHashValues: NonEmptyChain[RhoHash]
  ): F[Eta] =
    Sync[F]
      .delay(
        EtaCalculationArgs(
          previousEta,
          epoch,
          rhoNonceHashValues.toIterable
        ).digestMessages
      )
      .map(_.map(_.toByteArray))
      .flatMap(bytes => CryptoResources[F].blake2b256.useSync(_.hash(bytes*)))
      .map(ByteString.copyFrom)

}
case class EtaCalculationArgs(
    previousEta: Eta,
    epoch: Epoch,
    rhoNonceHashValues: Iterable[RhoHash]
) {

  def digestMessages: List[Bytes] =
    previousEta +:
      epoch.immutableBytes +:
      rhoNonceHashValues.toList
}
